guppylang-internals/src/guppylang_internals/definition/protocol.py

```diff
@@ -0,0 +1,149 @@
+        "a declaration"
+    )
+
+
+@dataclass(frozen=True)
+class RawProtocolDef(ProtocolDef, ParsableDef):
+    """A raw protocol definition that has not been parsed yet."""
+
+    python_class: type
+
+    def parse(self, globals: Globals, sources: SourceMap) -> "CheckedProtocolDef":
+        """Parses the raw class object into an AST and checks that it is well-formed."""
+        # Mostly copied from RawStructDef.parse, but only allowing function declarations
+        # in the body and allowing `Protocol` as a base class.
+        frame = DEF_STORE.frames[self.id]
+        cls_def = parse_py_class(self.python_class, frame, sources)
+        if cls_def.keywords:
+            raise GuppyError(UnexpectedError(cls_def.keywords[0], "keyword"))
+
+        # Look for generic parameters from Python 3.12 style syntax.
+        params = []
+        params_span: Span | None = None
+        if sys.version_info >= (3, 12):
+            if cls_def.type_params:
+                first, last = cls_def.type_params[0], cls_def.type_params[-1]
+                params_span = Span(to_span(first).start, to_span(last).end)
+                params = [
+                    parse_parameter(node, idx, globals)
+                    for idx, node in enumerate(cls_def.type_params)
+                ]
+
+        match cls_def.bases:
+            case []:
+                pass
+            # We allow `Generic[...]` to specify  parameters with the legacy syntax.
+            case [base] if elems := try_parse_generic_base(base):
+                # Complain if we already have Python 3.12 generic params
+                if params_span is not None:
+                    err: Error = RedundantParamsError(base, self.name)
+                    err.add_sub_diagnostic(RedundantParamsError.PrevSpec(params_span))
+                    raise GuppyError(err)
+                params = params_from_ast(elems, globals)
+            # Specifying `Protocol` is redundant but we allow it optionally.
+            case [base] if isinstance(base, ast.Name) and base.id == "Protocol":
+                pass
+            case bases:
+                err = UnsupportedError(bases[0], "Protocol inheritance", singular=True)
+                raise GuppyError(err)
+
+        func_defs: dict[str, DefId] = {}
+        for i, node in enumerate(cls_def.body):
+            match i, node:
+                # Docstrings are fine if they occur at the start.
+                case 0, ast.Expr(value=ast.Constant(value=v)) if isinstance(v, str):
+                    pass
+                # Ensure that all function definitions are Guppy declarations.
+                case _, ast.FunctionDef(name=name) as node:
+                    from guppylang.defs import GuppyDefinition
+
+                    v = getattr(self.python_class, name)
+                    if not isinstance(v, GuppyDefinition):
+                        raise GuppyError(NonGuppyMethodError(node, self.name, name))
```

Comment:
Do we actually want to require `@guppy.declare` annotations on protocol methods? We could also leave them unannotated and parse them here?